Add allow_rebase_history flag and fix codegen functions for multiple views

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32940 Change to only warn for custom Function
* #32839 Clean views created inside no_grad that are modified inplace
* **#32790 Add allow_rebase_history flag and fix codegen functions for multiple views**

Same as https://github.com/pytorch/pytorch/pull/31990 but without the first commit in the stack that is problematic for a lot of people.

Differential Revision: [D19814116](https://our.internmc.facebook.com/intern/diff/D19814116)

albanD committed:
Add allow_rebase_history flag and fix codegen functions for multiple views

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 9a93c942:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=9a93c9426bc8d980733085b9e14e2fa08f844619).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4374216) pytorch_windows_test1 (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92838465) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92838465))
<details>
<summary>
<code>RuntimeError: test_nn failed!</code>
</summary>

```
 
---------------------------------------------------------------------- 
Ran 1621 tests in 591.871s 
 
FAILED (errors=8, skipped=68, expected failures=4) 
Traceback (most recent call last): 
  File "run_test.py", line 457, in <module> 
    main() 
  File "run_test.py", line 450, in main 
    raise RuntimeError(message) 
RuntimeError: test_nn failed! 
 
(base) circleci@PACKER-5E29F737 C:\Users\circleci\project\test>if ERRORLEVEL 1 exit /b 1  
+ cleanup
+ retcode=1
+ set +x
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

albanD committed:
Update on "Add allow_rebase_history flag and fix codegen functions for multiple views"


Same as https://github.com/pytorch/pytorch/pull/32044 but without the first commit in the stack that is problematic for a lot of people.

[ghstack-poisoned]

albanD committed:
Update on "Add allow_rebase_history flag and fix codegen functions for multiple views"


Same as https://github.com/pytorch/pytorch/pull/32044 but without the first commit in the stack that is problematic for a lot of people.

[ghstack-poisoned]

albanD committed:
Update on "Add allow_rebase_history flag and fix codegen functions for multiple views"


Same as https://github.com/pytorch/pytorch/pull/32044 but without the first commit in the stack that is problematic for a lot of people.

[ghstack-poisoned]

albanD committed:
Update on "Add allow_rebase_history flag and fix codegen functions for multiple views"


Same as https://github.com/pytorch/pytorch/pull/32044 but without the first commit in the stack that is problematic for a lot of people.

[ghstack-poisoned]

ezyang commented:
Uhhh is the old PR description still up to date?

ezyang commented on tools/autograd/gen_autograd.py:
```diff
@@ -67,6 +65,20 @@
+#   backward as the multiple Nodes we create will force the gradient buffer to be created
+#   for each Node, instead of once in the original multi-output Node.
+# If you update this list, please update "test_multi_view_methods" in test_autograd.py
+PURE_VIEW_FUNCTIONS = ['split', 'split_with_sizes']
```
nit: make it a set?

Explain to me again why the default is that all multi-return view functions are non-pure, as opposed to pure? (The opposite seems to intuitively make more sense to me)

ezyang replied:
I was also expecting this to say something about unbind here (which morally should be in the list, but iirc you decided that you had to keep the old behavior first, right?) It's totally non-obvious as written here that the omission of unbind is intentional

albanD replied:
I will rework the comment to make it clearer that we only introduce this to ensure backward compatibility with functions that have a wrong behavior. But that we cannot change.

albanD replied:
Actually, because only the other commit introduces these functions as view, I'll just remove this from this PR.

ezyang commented on tools/autograd/derivatives.yaml:
```diff
@@ -94,6 +94,22 @@
+#           view operation, it can be added to PURE_VIEW_FUNCTIONS in gen_autograd.py. In that
+#           case, the outputs won't be created with allow_rebase_history=false but the backward
+#           specified in this file might NOT be used. See the note in gen_autograd.py for
+#           more details. (split for example)
```
Similarly, this comment doesn't make it clear why unbind doesn't satisfy the critera for "pure view functions".  It's only doing a view operation!

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -295,6 +295,13 @@ struct TORCH_API AutogradMeta : public c10::AutogradMetaInterface {
 ///     is updated.
+///   + if a single autograd Node returns multiple differentiable views, if any
+///     output is modified by an inplace operation, the autograd engine will make
+///     an equivalent graph (corresponding to the view operations) without using
```
editing suggestion: equivalent graph, where each output is treated as if it were produced by a distinct view operation. This discards the original (e.g., user provided) `grad_fn`

ezyang commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -97,23 +108,44 @@ template<typename... Args> inline variable_list flatten_tensor_args(Args&&... ar
 
 // See NOTE [ Autograd View Variables ] for details.
-inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable = true) {
+inline Tensor as_differentiable_view(const Tensor & base, Tensor tensor, bool allow_rebase_history) {
```
Duplication here is not great

albanD replied:
I wasn't sure how to handle the fact that differentiable views take the `allow_rebase_history` argument but not the non-differentiable ones.
You would prefer that I merge both back and add the `allow_rebase_history` to both (asserting that it is `true` for non differentiable views)?

ezyang replied:
Yes, I think that's my preference. If you want to deal with boolean blindness, make a little enum and pass that in as the input rather than just a boolean true or false.

ezyang commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -41,9 +41,20 @@ namespace torch { namespace autograd {
+    } else if (var.is_view()) {
+      // NB: is_view() ==> get_autograd_meta()
+      auto diff_view_meta = static_cast<DifferentiableViewMeta*>(impl::get_autograd_meta(var));
+      auto grad_fn = impl::grad_fn_unsafe(var);
```
But you only need to retrieve the grad fn if the check fails, is that right?

albanD replied:
This is an internal assert because this should never happen: `!var.is_leaf() ==> var.grad_fn_ != nullptr` according to the current definition of leaf.

ezyang commented:
Can we see before and after of the generated code?

albanD commented:
The comment in the previous PR is up to date. The only difference is that there is no whitelist here as it was the other PR in the stack that required the introduction of the whitelist. Also the link was to the wrong PR in the stack. I fixed it.

Old version:
```cpp
std::vector<Tensor> unbind(const Tensor & self, int64_t dim) {
  RECORD_FUNCTION("unbind", std::vector<c10::IValue>({self}), Node::peek_at_next_sequence_nr());
  auto& self_ = unpack(self, "self", 0);
  std::shared_ptr<UnbindBackward> grad_fn;
  if (compute_requires_grad( self )) {
    grad_fn = std::shared_ptr<UnbindBackward>(new UnbindBackward(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->dim = dim;
  }
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    op_name = jit::Symbol::fromQualString("aten::unbind");
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "dim", dim);
    tracer_state->graph->insertNode(node);
  
    jit::tracer::setTracingState(nullptr);
  }
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto tmp = ([&]() {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    return at::unbind(self_, dim);
  })();
  auto result = as_view(self, tmp, true);
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, result);
  }
  return result;
}
```
New version:
```cpp
std::vector<Tensor> unbind(const Tensor & self, int64_t dim) {
  RECORD_FUNCTION("unbind", std::vector<c10::IValue>({self}), Node::peek_at_next_sequence_nr());
  auto& self_ = unpack(self, "self", 0);
  std::shared_ptr<UnbindBackward> grad_fn;
  if (compute_requires_grad( self )) {
    grad_fn = std::shared_ptr<UnbindBackward>(new UnbindBackward(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->dim = dim;
  }
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    op_name = jit::Symbol::fromQualString("aten::unbind");
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "dim", dim);
    tracer_state->graph->insertNode(node);
  
    jit::tracer::setTracingState(nullptr);
  }
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto tmp = ([&]() {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    return at::unbind(self_, dim);
  })();
  auto result = as_differentiable_view(/* base */self, /* output */ tmp, /* allow_rebase_history */ false);
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, result);
  }
  return result;
}
```

Only the `as_view` line is changed to properly set the `allow_rebase_history` flag.


@ezyang following your comments, I did a big cleanup of the docs that were still assuming the first commit in the stack was here. This PR is much simpler actually it just adds the `allow_rebase_history=false` to **all** multi-view functions. And so now, instead of raising an internal assert, they raise a nice error when trying to modify a result inplace.

albanD committed:
Update on "Add allow_rebase_history flag and fix codegen functions for multiple views"


Same as https://github.com/pytorch/pytorch/pull/31990 but without the first commit in the stack that is problematic for a lot of people.

[ghstack-poisoned]

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -759,59 +762,38 @@ def declare_returned_variables():
+                return 'as_non_differentiable_view({}, {})'.format(view_info, call)
+            elif len(differentiable_output_vars) == 1:
+                # Single differentiable output (Tensor or Tensor[])
+                return_info = returns[0]
```
`assert len(returns) == 1` ?

albanD replied:
Right we should be using `return_info = differentiable_outputs[0]` here.

ezyang reviewed (approved):
okey dokey

albanD committed:
Update on "Add allow_rebase_history flag and fix codegen functions for multiple views"


Same as https://github.com/pytorch/pytorch/pull/31990 but without the first commit in the stack that is problematic for a lot of people.

[ghstack-poisoned]

albanD committed:
Update on "Add allow_rebase_history flag and fix codegen functions for multiple views"


Same as https://github.com/pytorch/pytorch/pull/31990 but without the first commit in the stack that is problematic for a lot of people.

Differential Revision: [D19814116](https://our.internmc.facebook.com/intern/diff/D19814116)

[ghstack-poisoned]

albanD committed:
Update on "Add allow_rebase_history flag and fix codegen functions for multiple views"


Same as https://github.com/pytorch/pytorch/pull/31990 but without the first commit in the stack that is problematic for a lot of people.

Differential Revision: [D19814116](https://our.internmc.facebook.com/intern/diff/D19814116)

[ghstack-poisoned]
